reactor3/src/main/java/com/linecorp/armeria/common/reactor3/RequestContextHooks.java

```diff
@@ -0,0 +1,259 @@
+     * The reactor {@link Publisher}s such as {@link Mono} and {@link Flux} will have the
+     * {@link RequestContext} which is in the {@link RequestContextStorage} when the {@link Publisher}s
+     * are created. Then, the {@link RequestContext} is propagated during the
+     * operations so that you can get the context using {@link RequestContext#current()}.
+     *
+     * <p>However, please note that {@link Mono#doOnCancel(Runnable)}, {@link Mono#doFinally(Consumer)},
+     * {@link Flux#doOnCancel(Runnable)} and {@link Flux#doFinally(Consumer)} will not propagate the context.
+     *
+     * <p>Also, note that this method does not have any relevance to {@link Context} API.
+     */
+    public static synchronized void enable() {
+        if (enabled) {
+            return;
+        }
+        Hooks.onEachOperator(ON_EACH_OPERATOR_HOOK_KEY, source -> {
+            if (source instanceof ContextHolder) {
+                return source;
+            }
+
+            if (source instanceof Scannable) {
+                final Object parent = ((Scannable) source).scanUnsafe(Attr.PARENT);
+                if (parent instanceof ContextHolder) {
+                    return makeContextAware(source, ((ContextHolder) parent).context());
+                }
+            }
+            return RequestContext.mapCurrent(requestContext -> makeContextAware(source, requestContext),
+                                             () -> source);
+        });
+
+        Hooks.onLastOperator(ON_LAST_OPERATOR_HOOK_KEY, source -> {
+            if (source instanceof ContextHolder) {
+                return source;
+            }
+
+            if (source instanceof Scannable) {
+                final Object parent = ((Scannable) source).scanUnsafe(Attr.PARENT);
+                if (parent instanceof ContextHolder) {
+                    return makeContextAware(source, ((ContextHolder) parent).context());
+                }
+            }
+            return source;
+        });
+
+        enabled = true;
+    }
+
+    /**
+     * Disables {@link RequestContext} during Reactor operations.
+     */
+    public static synchronized void disable() {
+        if (!enabled) {
+            return;
+        }
+        Hooks.resetOnEachOperator(ON_EACH_OPERATOR_HOOK_KEY);
+        Hooks.resetOnLastOperator(ON_LAST_OPERATOR_HOOK_KEY);
+        enabled = false;
+    }
+
+    private static Publisher<Object> makeContextAware(Publisher<Object> source, RequestContext ctx) {
+        if (source instanceof Mono) {
+            return new ContextAwareMono((Mono<Object>) source, ctx);
+        }
+        if (source instanceof ConnectableFlux) {
+            return new ContextAwareConnectableFlux((ConnectableFlux<Object>) source, ctx);
+        }
+        if (source instanceof ParallelFlux) {
+            // TODO(minwoox) Support ParallelFlux after
+            //               https://github.com/reactor/reactor-core/issues/2328 is addressed.
+            return source;
```

Comment:
Should we log a warning message once when we reach here for potentially less confusion?

Reply:
Added. 😉